lib/schema.js

```diff
@@ -135,97 +102,100 @@ class Schema {
     this.indexes = {};
     this.references = new Set();
     this.relations = new Set();
-    this.validate = defs.validate || null;
-    this.format = defs.format || null;
-    this.parse = defs.parse || null;
-    this.serialize = defs.serialize || null;
-    this.preprocess(defs);
+    this.validate = setValidate(raw.validate);
+    this.format = raw.format || (() => {});
+    this.parse = raw.parse || (() => {});
+    this.serialize = raw.serialize || (() => {});
+    this.fields = this.preprocess(raw);
   }
 
-  preprocess(defs) {
-    const keys = Object.keys(defs);
-    const first = keys[0];
-    if (isFirstUpper(first)) {
-      const metadata = defs[keys.shift()];
-      this.kind = toLowerCamel(first);
-      this.scope = metadata.scope || 'application';
-      this.store = metadata.store || 'persistent';
-      this.allow = metadata.allow || 'write';
-      if (metadata.schema && metadata.fields) {
-        this.parent = metadata.schema;
-        const parent = this.findReference(this.parent);
-        for (const key of metadata.fields) {
-          defs[key] = parent.fields[key];
-          keys.push(key);
-        }
-      }
+  preprocess(raw) {
+    const { name, def, kind, metadata } = this.preprocessType(raw);
+    if (name !== 'schema') {
+      const type = this.findType(name);
+      if (!type) throw new Error(`Unknown type ${type}`);
+      const long = type.toLong(def, Schema, this);
+      this.kind = this.kind || type.kind;
+      return { value: long };
     }
+    this.kind = kind;
+    if (!this.findKind(kind)) throw new Error('Given Kind is invalid');
+    this.applyMetadata(metadata);
+    const extra = kind === 'projection' ? this.projection(metadata) : {};
```

Comment:
Nit: We should consider some "streamlined" projection handling to avoid such `if projection else everything else` checks